### PR TITLE
Adding undersores for class dict instances

### PR DESCRIPTION
**Pull request by JustusAdam**
First of hopefully two pull requests to address the issue in #486. 

This changes the generation function for class instance dicts.

Class instances

```hs
class MyClass a b c where

instance MyClass Tt Aa b where
instance MyClass Tt A ab where
```

Would previously both result in the instance dict named `$fMyClassTtAab`.
To reduce name clashes and make instances more readable we introduce a `_` as separator for individual type arguments resulting in the two dicts `$fMyClass_Tt_Aa_b` and `$fMyClass_Tt_A_ab`.

## Dicussion

Considering that we need further changes to address the case-sensitivity issue do we want to escape literal `_`?

Aka should 

`MyClass T_t` be `$fMyClass_T_t` (no escaping)
or 
`MyClass T_t` be `$fMyClass_T__t` (`_` escaped to `__`)?

### CODE REVIEW

**Comment by rahulmutt** [comment]
The build failed with this:
```
java.lang.NoSuchMethodException: base.java.Exception.$fExceptionJException()
	at java.lang.Class.getMethod(Class.java:1786)
	at eta.runtime.stg.Closures.loadClosure(Closures.java:63)
	at eta.runtime.stg.Closures.<clinit>(Closures.java:50)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at Verify.main(Verify.java:39)
```
Can you also change `eta.runtime.stg.Closures` to reference `$fException_JException` instead of `$fExceptionJException`? Also run `tests/verify/verify.sh` to ensure that this change works.

**Comment by rahulmutt** [comment]
And please do escaping because `_` is an artifact of the encoding we're doing and `_` itself should be able to be distinguished from the separator `_`.

File: compiler/ETA/TypeCheck/TcEnv.hs [hunk 1]
```diff
@@ -784,12 +785,26 @@ newDFunName :: Class -> [Type] -> SrcSpan -> TcM Name
...
-        ; let info_string = occNameString (getOccName clas) ++
-                            concatMap (occNameString.getDFunTyKey) tys
-        ; dfun_occ <- chooseUniqueOccTc (mkDFunOcc info_string is_boot)
+        ; dfun_occ <- chooseUniqueOccTc $ mkDFunOcc (mkInfoString clas tys) is_boot
         ; newGlobalBinder mod dfun_occ loc }
 
 {-
+It may be advantageous at some point to build a string for a type which actually
+encompasses the structure of the type like parethesizing.
+Even in the new mkInfoString something like `a (b c)` and `(a b) c` will result
```
Comment (rahulmutt) [inline compiler/ETA/TypeCheck/TcEnv.hs]: You can handle parenthesis with Z-encoding, but I'm not sure it's worth it at this point. We can do that if it seriously becomes an issue.

**Comment by JustusAdam** [comment]
Okay, interesting. For me this morning for some reason the build worked and also the verify script. It didn't a few days back when I tried it. I suppose something is wrong with my test setup, I'll have to figure that out.

I also considered some kind of encoding for parentheses, but I couldn't think of one that I would have deemed readable.

**Comment by rahulmutt** [comment]
You can run `./cleaninstall.sh` to ensure that the old version is cleanly removed. Another issue may be if you did the binary installation first, in which case `etlas` would be using your downloaded binary and not the source installed `eta`.

Yup, as I said, nothing to worry about (in regards to the parenthesis).

**Comment by rahulmutt** [comment]
Hey @JustusAdam, let me know if you have time to fix the naming change in the runtime as well. If not, I'll merge this and immediately push the fix.

**Comment by JustusAdam** [comment]
Oh if you want you can fix it. In fact you can fix it on this branch. I have enabled `allow edits from maintainers` so you should be able to push directly into this PR.
I might get around to fixing it today but I am not sure.

**Comment by rahulmutt** [comment]
Merged, thanks for bringing this in!

**Pull request merged**